### PR TITLE
Update rust toolchain to 2023-04-29

### LINKED ISSUE #2451: Update rust toolchain version to nightly-2023-04-30

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

Update toolchain to 2023-04-29. Relevant PRs that required changes:

Removal of WithOptConstParam: https://github.com/rust-lang/rust/pull/96840
Moving of IndexVec into its own module: https://github.com/rust-lang/rust/pull/110539
Hashes use their own type instead of `u64`: https://github.com/rust-lang/rust/pull/110083
New `OffsetOf` variant in `rustc_middle::mir::NullOp`: https://github.com/rust-lang/rust/pull/106934
Lowering of `offset` intrinsic to `BinOp::Offset`: https://github.com/rust-lang/rust/pull/110822

Also had to add an `isize` suffix to the second argument to the `offset` intrinsic due to https://github.com/rust-lang/rust/issues/111699.

### Resolved issues:

Towards #2451 

### Related RFC:

Optional #ISSUE-NUMBER.

### Testing:

* How is this change tested? Current regressions

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by zhassan-aws** [comment]
Keeping this as a draft because there are still a few failures in the `expected` test suite. Also, the `codegen_firecracker.sh` regression doesn't terminate after running for more than 10 minutes.

**Comment by zhassan-aws** [comment]
All regressions now pass except for two perf regressions that are running out of memory 😞 
```
tests/perf/btreeset/insert_multi
tests/perf/btreeset/insert_same
```

**Comment by zhassan-aws** [comment]
On one of the BTreeSet performance tests (`tests/perf/btreeset/insert_any`), the number of nondets in CBMC's `--program_only` output increased significantly with the toolchain update:
```
$ grep -c nondet before.po 
117
$ grep -c nondet after.po 
692
```
The size of the program expression also increased > 5X:
```
$ grep "program expression" before.po 
size of program expression: 15889 steps
$ grep "program expression" after.po 
size of program expression: 87287 steps
```
@tautschnig is there a way to pinpoint what is causing the difference?

**Comment by zhassan-aws** [comment]
I narrowed it down to a change that happened between the 2023-04-29 and 2023-04-30 nightly toolchains: just changing the toolchain to 2023-04-29 in this PR restores performance on those btreeset tests.

**Comment by zhassan-aws** [comment]
For the sake of making progress on the toolchain updates, I've backtracked the toolchain date by 1 day to 2023-04-29 to avoid the performance regression that occurs with the 2023-04-30 toolchain.

The PR is now ready for review.

**Review by celinval — COMMENTED** [review]
Thanks Zyad! I just have questions about the changes to offset implementation.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 1]
```diff
@@ -317,7 +318,32 @@ impl<'tcx> GotocCtx<'tcx> {
             BinOp::Offset => {
                 let ce1 = self.codegen_operand(e1);
                 let ce2 = self.codegen_operand(e2);
-                ce1.plus(ce2)
+
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:321]: Yes, I had to change this. Wtih https://github.com/rust-lang/rust/pull/110822, the `offset` intrinsic is now lowered to `BinOp::Offset`, so I had to move the checks we do for the `offset` intrinsic here.

[the remark this answers is not in the document]
Comment (celinval) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:321]: Yeah, I looked at the rustc change. Was this logic in the intrinsic shared by other operations or should we remove it?

File: tests/expected/offset-from-bytes-overflow/main.rs [hunk 2]
```diff
@@ -9,7 +9,7 @@ use std::convert::TryInto;
 fn main() {
     let v: &[u128] = &[0; 10];
     let v_0: *const u128 = &v[0];
-    let high_offset = usize::MAX / (std::mem::size_of::<u128>() * 2);
+    let high_offset = usize::MAX / (std::mem::size_of::<u128>());
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply tests/expected/offset-from-bytes-overflow/main.rs:12]: Not really. The current expected failure stopped failing with the changes, which I think is correct, so I removed the `* 2` to restore the failure.

This test is actually meant to check overflow in the `offset_from` method, but what is actually failing is the check in the `add` call which blocks further execution, and so we don't reach the `offset_from` line. I'm not sure if there's an easy way to trigger the failure that this test is looking for.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:321]: The `codegen_offset` function is still used by the `arith_offset` intrinsic.

File: tests/expected/offset-from-bytes-overflow/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply tests/expected/offset-from-bytes-overflow/main.rs:12]: Here's why I think the failure with the `* 2` included was flaky:

- `usize::MAX / (std::mem::size_of::<u128>() * 2` is `18446744073709551615 / 32` which is `576460752303423487`. To compute the offset for the `add`, we multiply this number by the bytes in a `u128` (16), so this gives us `9223372036854775792` which is less than `isize::MAX` (`9223372036854775807`). Now depending on the pointer value that CBMC picks for `v_0`, the `add` may or may not overflow `isize`. Thus, by removing the `* 2`, the add operation **must** overflow. Does this make sense?

**Review by celinval — APPROVED** [review]

**Pull request merged**